src/terminal/adapter/adaptDispatch.cpp

```diff
@@ -1176,9 +1209,19 @@ bool AdaptDispatch::ReverseAttributesRectangularArea(const VTInt top, const VTIn
             // rendition bits. But note that this shouldn't be triggered by an
             // empty attribute list, so we explicitly exclude that case in
             // the empty check above.
-            if (attrs.at(i).value_or(0) == 0)
+            const auto param = attrs.at(i).value_or(0);
+            if (param == 0)
             {
-                reverseMask ^= CharacterAttributes::Rendition;
+                reverseMask ^= CharacterAttributes::Rendition & ~CharacterAttributes::UnderlineStyle; // underline bits do not take part in reverseMask
+                reverseUnderline = !reverseUnderline;
+                i++;
+            }
+            else if (param == 4 || param == 21)
+            {
+                if (!attrs.hasSubParamsFor(i)) // ignore the 4:x format
+                {
+                    reverseUnderline = !reverseUnderline;
+                }
                 i++;
             }
```

Comment:
I'm sorry I haven't have time to look at this properly, so this is just a drive-by comment, but if it were up to me I'd drop all this special case handling for `reverseUnderline`, since it's not actually working meaningfully for the different underline styles, and it seems like unnecessary overhead.

If you just let the underline style bits XOR the same way as everything else, that should work fine for the typical use case. For example, if you've got some content on the screen which has a wavy underline, and you want to toggle that off and on again, you can just use `DECRARA` with a `4:3` attribute.

You're going to get some weird results if you're applying that to content that has some other style of underline, but that's fine - there's no expectation for that to work in any particular way. And this would at least be reversible, i.e. if you reapply the `DECRARA` a second time, you get back what you started with, which is quite important.

The only thing possibly worth tweaking is the `param == 0` case. Maybe just filter out the top two underline style bits, and that way it should work "correctly" as long as you stick to single underlines. The 0 param is a weird case anyway, and I'm not positive we've got it right as it is now, so I wouldn't worry about it too much.

Reply:
The changes for `DECRARA` is somewhat opinionated TBH 😅 and I'd like to explain it.

`reverseUnderline` really has two jobs:

- Help us differentiate `DECRARA` and `DECCARA` within the code which is shared by both, since DECCARA already works, and only DECRARA needed the modification.
- Treat underline style as a boolean rather than 3 bit value.

So it's really a two things done at once. I wish we could split the implementation for both these ops, since they sound similar, but has a quite different way of handling attributes. `DECCARA` decides the *final* state of the attr and works with almost all of the SGR parameters, while OTH, `DECRARA` only decides which attribute needs to be reversed (final state is only known after looking at the current attr), and works with a very restricted set of SGR parameters. Given the current implementation, having a way out with `reverseUnderline` felt like a right approach to me, and it doesn't require any changes to how we handle `DECCARA`.

> you've got some content on the screen which has a wavy underline, and you want to toggle that off and on again, you can just use DECRARA with a 4:3 attribute.

This can get really wierd if the sequence includes multiple 4:x parameters (eg. `...4:3;4:2...`), since we accumulate the reverseMask, and only use the final form of it.

We should probably ignore 4:x form in DECRARA, so that we don't motivate such usage (unless there's some consensus around it else where). Let's keep only two state of an underline, we either reverse it or not reverse it, nothing In-between. Easy to 'manage'.

> Maybe just filter out the top two underline style bits

I'm not sure if this is any better. Since we still need to (not reverse but) turn off higher two bits within the current attr, for which we need some form of indication that we're in DECRARA and need to do this specific thing within the shared code.

> The 0 param is a weird case anyway, and I'm not positive we've got it right as it is now, so I wouldn't worry about it too much.

I wouldn't like to make it more wierd than it is already 🙂. As underline style required some changes in it, I did it so it doesn't break in odd ways and also behave the same way as it used to.

> if you reapply the DECRARA a second time, you get back what you started with, which is quite important.

I think DECRARA is implemented by so few terminals, I'd be surprise to see an application using extended underline style in conjunction with DECRARA 😅. I'd be happy to be proofed wrong here 😄

Reply:
Regardless, I'm open to suggestions! 😀

Reply:
> Help us differentiate DECRARA and DECCARA within the code which is shared by both, since DECCARA already works, and only DECRARA needed the modification.

But with what I was suggesting they still wouldn't need to be handled differently.

> Treat underline style as a boolean rather than 3 bit value.

But that's not exactly what it's doing. Single underline toggles all underline styles. Double underline is an alias for single underline. And the other styles do nothing. That inconsistency feels kind of broken to me.

In addition to that, you can't use `DECRARA` to set any underline style other than single, and toggling it on and off on cells with other underline styles will destroy the original style.

And while I don't think the exact behavior is that important, it bothers me that the `reverseUnderline` implementation is adding complexity and overhead to both `DECRARA` and `DECCARA` to produce a result which isn't that great.

> This can get really wierd if the sequence includes multiple 4:x parameters (eg. ...4:3;4:2...), since we accumulate the reverseMask, and only use the final form of it.

But that doesn't seem like a big deal to me. Doing something weird and getting weird results is fine. It's when you do something simple and get weird results that I have a problem with. Like `DECRARA` with 21 producing a single underline when you'd expect a double underline.

> Since we still need to (not reverse but) turn off higher two bits within the current attr

No. That's the point I'm trying to make. I don't think those edge cases matter. If you apply a single underline `DECRARA` to a cell with a wavy underline and it turns it into a double underline, that's fine. That's just undefined behavior - there's no expectation for how that should work.

> I did it so it doesn't break in odd ways and also behave the same way as it used to.

But it doesn't behave the same way that it used to, and it does break in odd ways.

> I think DECRARA is implemented by so few terminals, I'd be surprise to see an application using extended underline style in conjunction with DECRARA

It's not using extended underline styles as parameters to `DECRARA` that concerns me. It's applying a single underline `DECRARA` over existing content that might have extended underline styles. As I mentioned above, I don't particularly care what that looks like, but I do care that the operation is reversible.

An example use case would be something like an editor highlighting the focused row by applying `DECRARA` to that row with an underline style. It's an easy way to highlight something without having to redraw all of the content. And you can remove the highlight by simply reapplying the same `DECRARA` sequence. That's exactly the point of `DECRARA`, but it doesn't work if the operation isn't reversible.

> I'd be surprise to see an application using extended underline style in conjunction with DECRARA

I would too, but I wouldn't be surprised to find an app using `DECRARA` for highlighting the way I described above, and also having support for users to configure their own styling. If the user decides to style spelling errors with a wavy underline, then the rendering is going to break.

That said, I do agree this usage is going to be rare, so it's not a big deal. But I also don't see any obvious use cases for the current implementation that justifies the additional complexity.

Anyway, I'll leave it at that. If you want to stick with it as it is, that's fine. I just wanted to offer my opinion.

Reply:
> Single underline toggles all underline styles. Double underline is an alias for single underline. And the other styles do nothing. That inconsistency feels kind of broken to me.

DECRARA as per the Standard, only support the parameter 4, and parameter 21 isn't supported. Because we're going to support underline style, I've extended parameter 4 to also toggle all underline styles (except doubly underlined, more on this later). My interpretation of the parameter `4` is that it toggles underlines, since any application that is reversing the underline either wanted to change:

1. Singly underline to no-underline.
2. No-underline to singly-underline.

Now, I've extended this to:

1. *Any underline* to no-underline.
2. No-underline to singly underline.

Any underline in (1), because I thought that's what most application might be expecting. I could be so wrong here.

DECRARA supporting *SGR 4:x* is not something that I intended in this PR. It needs separate discussion on how that would work, specially, since I couldn't find any application that I could refer.

> Double underline is an alias for single underline.

Ahh, I didn't intend that. I'll fix it in no time 😅. But do we really need it? parameter 21 isn't supported in the standard DECRARA.

Reply:
> It's not using extended underline styles as parameters to DECRARA that concerns me. It's applying a single underline DECRARA over existing content that might have extended underline styles. As I mentioned above, I don't particularly care what that looks like, but I do care that the operation is reversible.

If it needs to be reversible then we need to support SGR 4:x in DECRARA which I mentioned is outside of  this PR. Otherwise, we need to track the last underline style we reversed from, which shouldn't be much difficult to do.

> An example use case would be something like an editor highlighting the focused row by applying DECRARA to that row with an underline style. It's an easy way to highlight something without having to redraw all of the content.

I'm curious why DECRARA is preferred over DECCARA for this. DECCARA supports *SGR 4:x* and is supported by most terminals.

> And you can remove the highlight by simply reapplying the same DECRARA sequence. That's exactly the point of DECRARA, but it doesn't work if the operation isn't reversible.

This is an interesting case. Not sure how this should be done while adhering to the standard. Track underline style that was reversed? or allow application to specify which underline style should be toggled using `SGR 4:x` (this is probably a future enhancement, and requires a discussion on it). The right thing for now would be to consider that DECRARA *doesn't* support `SGR 4:x` and ignore the sequence. `_CanSeqAcceptSubParam(...)` in `OutputStateMachineEngine` needs to be updated.

Reply:
> DECRARA as per the Standard, only support the parameter 4, and parameter 21 isn't supported.

That's because none of the DEC terminals supported 21 as an SGR attribute. There are a whole bunch of renditions attributes that we support in `DECRARA` which they never supported in `SGR` or `DECRARA`.

That said, the underline styles are different, in that they potentially cancel each other out. So I wouldn't really care if we filtered out everything except the basic 4 attribute. But it just seems easier to accept everything and XOR the bits as they are. It's less work, and does actually provide some additional functionality if anyone wanted it.

> Now, I've extended this to:
>
> 1. Any underline to no-underline.
> 2. No-underline to singly underline.
>
> Any underline in (1), because I thought that's what most application might be expecting.

I wouldn't have an objection to (1) if it were reversible, but it's not. That's the most important aspect of `DECRARA`. If you wanted to convert any underline to no-underline without needing to get the original back again, you could just use `DECCARA`.

> If it needs to be reversible then we need to support SGR 4:x in DECRARA

As long as you stick to XORing bits, and nothing thing else, it's _guaranteed_ to be reversible, regardless of which underline styles you accept. You may not like the way it looks when you XOR a single underline with some other style of underline, but I just don't think that's worth worrying about, at least initially. Unless you have a specific use case in mind?

> I'm curious why DECRARA would be preferred over DECCARA for this. DECCARA supports SGR 4:x and is supported by most terminals.

Again, because it's not reversible. That's the _key_ feature. If the row that you're highlighting already has some elements that are underlined, and you overwrite the whole row with `DECCARA`, it's now impossible to get the row back to what it was. To remove the underline that you've added with `DECCARA`, you have to remove it from everything.

> The right thing for now would be to consider that DECRARA doesn't support SGR 4:x and ignore the sequence. _CanSeqAcceptSubParam(...) in OutputStateMachineEngine needs to be updated.

If you want to drop 4:x support, it's probably easier to just filter it out in the loop where you're building the reverseMask. But again this is neither necessary, nor sufficient to solve the reversible problem. As long as you insist on retaining the current "boolean underline" implementation, the operation is not going to be reversible.

Just comparing the pros and cons of the different approaches. For a simple XOR mask of the underline style bits as they are, you've got:

PRO
1. It's guaranteed to be reversible.
2. It can be used with any underline style. 
3. It adds no overhead to the existing code.

CON
1. Mixing one underline style with another can produce weird results.

---

For the boolean underline approach you've got:

PRO
1. Mixing one underline style with another is less weird.

CON
1. It's not fully reversible.
2. You can't use any style other than single underline. 
3. It introduces additional complexity to both `DECCARA` and `DECRARA`.

Am I'm missing any other advantages? Because otherwise I just don't get the appeal of this approach.

Reply:
I'll get this out of the way first: I'm cool with the XOR approach, and accidentally having multiple underline styles active at the same time. The renderer can choose which one to apply, or apply them all. It might look ugly. It might look strange.

As long as `DECRARA` doesn't turn on _all of the underlines at the same time_ for any cells that don't have an underline. That would be wacky I guess?

---

There's some interesting behavior to consider, which is very point in time[^1].

If you have multiple underline styles active at the same time (which this would allow for, in a way that SGR `4:x` does not), ConPTY will have to send all of them or choose one of them.

If it sends all of them:

- The application on the receiving end will treat it as a sequence of _overrides_
- Only one will be applied in the end due to how SGR `4:x` works.

If it sends one of them:

- Only one will be applied in the end.
- I do not know why I made this a bulletted list.

Of course, flipping it back over with `DECRARA` would return to the original one.

[^1]: Only while conhost handles `DECRARA` and doesn't tell the ConPTY host about it.

Reply:
> If you have multiple underline styles active at the same time

@DHowett You can't have multiple styles at the same time since we're storing all 6 options in a 3-bit value (000 = off, 001 = single, 010 = double, 011 = curly, etc). This is also the source of weirdness with the XOR approach. If you `DECRARA` a single underline (001) over a cell with a curly underline (011), the result is going to be a double underline (010).

> As long as `DECRARA` doesn't turn on _all of the underlines at the same time_ for any cells that don't have an underline. That would be wacky I guess?

That's kind of possible with `DECRARA` parameter 0, which is supposed to reverse all attributes at the same time. But that's already extremely wacky considering it turns on all the gridlines, italics, blink, strike-out, etc. 

If we turned on _all the underline styles_ here, that would essentially be setting all 3 bits, which isn't technically anything, so my recommendation for param 0 was to restrict it to just the single underline bit. And I've been thinking we should possibly be filtering out many of the other attribute bits here as well, but that's a discussion for another time. This is very much an edge case anyway.